### DISCUSSION TITLE
Get Renovate to work with GitLab NPM Package registry

### DISCUSSION THREAD

**Discussion by jume-dev** [renovatebot/renovate#10049] [Request Help] [answered]
Hello there,
first of all, as this is my first discussion: Thank you for this tool, we use this for around 2 years now and are very pleased with it.
Sadly we now encounter a problem we can't seem to solve.

We switched recently from a public reachable Nexus host to the Gitlab NPM Package registry which now adds some authorization to our updating process.
We tried following all the documentation but we couldn't get it to work. I hope anyone can share some insights on that.

So, we have a repository from where we run scheduled pipelines over the night. This repo contains a `.gitlab-ci.yml` with the following:
```
include:
  - project: 'renovate-bot/renovate-runner'
    file: '/templates/renovate.gitlab-ci.yml'

renovate:
  only:
    - schedules
    - web
```

We also have a default renovate.json in here, which the other projects extend:
```
{
  "commitMessagePrefix": "[Dependency Update]",
  "extends": ["config:base"],
  "gitLabAutomerge": true,
  "labels": ["dependencies"],
  "packageRules": [
    {
      "automerge": false,
      "matchUpdateTypes": ["major"]
    },
    {
      "automerge": true,
      "excludePackageNames": ["vuetify"],
      "matchUpdateTypes": ["minor", "patch", "pin", "digest"],
      "groupName": "non-major dependencies",
      "groupSlug": "non-major-dependencies"
    },
    {
      "automerge": false,
      "matchPackageNames": ["vuetify"],
      "matchUpdateTypes": ["minor", "patch", "pin", "digest"]
    }
  ]
}
```

We provide 3 environment variables to the scheduled pipelines:
GITHUB_COM_TOKEN, RENOVATE_TOKEN and RENOVATE_EXTRA_FLAGS, which contains only the autodiscover stuff for finding our projects.

The RENOVATE_TOKEN is a PAT of our bot account with these scope `read_registry, read_user, write_repository and api`.

The individual projects contain an `.npmrc` file with the registry definition for our namespace like this:
```
@namespace:registry=https://gitlab.com/api/v4/projects/<project_id>/packages/npm/
```

We downloaded the log json file from the artifacts and it show, that it gets a 404 for the project id, which indicates that it isn't authenticated and we can see in the headers logged, that there isn't any token attached.

We assumed the RENOVATE_TOKEN would be enough for this but I guess we need to somehow tell renovate to use this for the registry API too.

Best regards

**Comment by viceice** [comment, marked as answer]
You need to configure hostrules like:

renovate runner repo *config.js*
```js

const projectIds = [ '123', '456' ];

const npmBaseRule = {
  hostType: 'npm',
  token: process.env.RENOVATE_TOKEN,
};

module.exports = {
  hostRules: [
    // other rules
    ...projectIds.map(id => ({...npmBaseRule, matchHost: `https://gitlab.com/api/v4/projects/${id}/packages/npm/`})),
  ]
}
```

https://docs.renovatebot.com/private-npm-modules/#add-hostrule-to-bots-config

**Reply by jume-dev** [reply]
Ah, thanks, I see. I guess we were confused with the config.js, because we tried to add the hostRules to the renovate.json

I'll try that and report later, if it worked.

**Reply by viceice** [reply]
another option is to use secrets

```js
module.exports = {
    secrets: {
        RENOVATE_TOKEN: process.env.RENOVATE_TOKEN,
    }
};
```


and then in target repo `renovate.json`

```json
{
  "hostRules": [
    {
      "hostType": "npm",
      "token": "{{ secrets.RENOVATE_TOKEN }}",
      "matchHost": "https://gitlab.com/api/v4/projects/<projectid>/packages/npm/"
    }
  ]
}
```

**Reply by jume-dev** [reply]
The first approach works, the second did not. 

It does seem to ignore the hostRules in the renovate.json somehow as the combinedConfig logstep doesn't contain it, but with the first approach it does.

But thanks anyway, It works now :)

**Reply by viceice** [reply]
What is the error with secrets?

**Reply by jume-dev** [reply]
The same as in the initial, it simply doesn't work. The hostRules from the renovate.json seem to be ignored for some reason

**Reply by viceice** [reply]
https://docs.renovatebot.com/self-hosted-configuration/#printconfig

Can you post the full config sanitized. Please only mask urls partly, so we can see the structure and if they match your registry setting in npmrc

**Reply by jume-dev** [reply]
Alright, here is the full log for a run on 1 repo, I thought that would be more helpful:
[log.json.txt](https://github.com/renovatebot/renovate/files/6522099/log.json.txt)
